### PR TITLE
Numerous changes to content app to better handle multiple content DBs

### PR DESCRIPTION
**Pull request by jamalex**
Note: This PR is targeting the `content_db_integration` branch, which merges together https://github.com/learningequality/kolibri/pull/188 and https://github.com/learningequality/kolibri/pull/196, with the latest master. Once this is merged, and those PRs have been reviewed, we can merge `content_db_integration` to master.

## Summary

- Added middleware to activate the content DB matching the `channel_id` component of the URL, removing need for `with ...` etc within any of the Django REST Framework machinery.
- Added base `ContentDatabaseModel` model which should be inherited by all models that live in the content database; this makes the logic for determining where to look for a model explicit. `ChannelMetadata` is split into `ChannelMetadata` (inheriting from `ContentDatabaseModel`) and `ChannelMetadataCache`, which does not, and lives in the main (default) database.
- Removed `rest_framework_nested`, and simplified the logic for the endpoints, relation, and reversals.
- Simplified some serializers into more standard model serializers.
- Updated tests and test fixtures, mostly to get them working fully with parallel changes (around primary keys, IDs, etc)
- Deleted content migrations and regenerated them, as changing the primary key of a model involved in M2M relations is not handled well. We will delete them one more time and regenerate after finalizing the IDs as described below.
- Simplified the pagination on `ContentNodeViewset` to just use the standard pattern (now that we don't need to pass channel as context)

## Discussion points for further iteration

- The content/api.py file contains the "interface for other apps to the content DB backend"; most of the functions in there seem like duplicates/aliases of the methods on ContentNode; as they return ContentNode objects anyway, and hence are providing direct access to all the methods on ContentNode, perhaps we should just add those of these that aren't simply aliases as helper methods on ContentNode. Also, a standard DRF convention is to use an app's api.py for things like DRF ViewSets, so perhaps we can repurpose this file.
- A lot of the detail_route's on ContentNodeViewset seem unnecessary (either not likely to be used, or better implemented as filters)
- The primary key for ContentNode is now called `instance_id` -- let's just rename it to `id`, to avoid complexity?
- Let's revisit the format of IDs across the content app (and hence also in the content curation server). It seems like it might make sense to use UUIDs for all non-fixture models (i.e. not license, language, etc -- but all user-generated ones)
- Let's decided how we want to handle related model references; for some (especially non-editable data, like License) we should probably just embed the full data. For non-embedded data, where we might want to request the related data, does it make more sense to use URLs or primary keys? I like URLs for the browsable API -- simply being able to click to navigate to the related model -- but for consumption by our JS code (and our Resource/Model/Collection stuff), it seems as if foreign keys may be more useful (or perhaps along with a resource name identifier, to enable reverse lookups and finding the right Resource on the client side?).
- sync_channelmetadata doesn't seem to belong in utils/validate.py -- let's find a better home for it

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -14,7 +14,8 @@ exclude = kolibri/*/migrations/*
 [isort]
 atomic = true
 multi_line_output = 5
-line_length = 79
+line_length = 160
+indent='    '
```
Comment (jamalex) [inline setup.cfg:18]: This is a fly-by fix that solves a problem I was having with precommit hooks -- the [isort] code was reducing indents to 2 spaces, and then flake8 was complaining about indents; this removes the catch-22.

File: kolibri/content/middleware.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (jamalex) [inline kolibri/content/middleware.py:1]: This middleware was what allowed most of the other simplifications to occur. Now, anything that happens in the context of a request to a URL with a `channel_id` component will automatically reference the correct DB.

**Comment by jamalex** [comment]
Flagged @66eli77 as this touches a lot of his code, @rtibbles so that we can look at some of the "discussion questions" around how this interacts with the frontend persistence layer, and @aronasorman as it interacts with the content DB export stuff.

**Comment by jamalex** [comment]
Added some additional tweaks and tests to boost content app test coverage to near 100%, and fixed https://github.com/learningequality/kolibri/issues/186 in the process.

**Comment by aronasorman** [comment]
Can you upgrade our django version as well, to at least 1.9.7, and ideally 1.10? There were some bugs that were fixed in the more recent versions, that would benefit us all and stop us from `:headdesk:`ing!

**Comment by jamalex** [comment]
> Can you upgrade our django version as well, to at least 1.9.7, and ideally 1.10?

This branch already has it upgraded to 1.9.7 (from your PR that was integrated here). I can upgrade to 1.10.

**Comment by jamalex** [comment]
TODO:

[ ] Remove api.py code (moving some to the models/querysets)
[ ] Remove unnecessary detail_route's on ContentNodeViewset
[x] Rename *_id pk's to just "id"
[x] Make all id's of non-fixture models UUIDs
[ ] Turn all non-embedded related model lookups into PK's (not URL's)
[ ] Move sync_channelmetadata out of validate.py
[ ] Rename urls.py to api_urls.py and mount at /api/ from root urls.py
[ ] Upgrade Django to 1.10

**Comment by jamalex** [comment]
@66eli77 @aronasorman @rtibbles I applied a lot of :fire: in https://github.com/learningequality/kolibri/pull/201/commits/4cfdb4aa8f8803700ff9ff04f70edb6fac68ffc3, but everything can now be done directly through contentnode filters. I also removed the "through" tables for related/prereq M2M relations, since we weren't adding any custom info in there, and it interfered with making "related" symmetric (and hence made the filter queries asymmetric -- could only query one direction -- which was weird).

**Comment by jamalex** [comment]
Not sure where the following travis failure is coming from; don't get it locally:

```
INFO     Automatically enabling applications.
Traceback (most recent call last):
  File ".tox/py2.7/bin/kolibri", line 9, in <module>
    load_entry_point('kolibri', 'console_scripts', 'kolibri')()
  File "/home/travis/build/learningequality/kolibri/kolibri/utils/cli.py", line 271, in main
    plugin(plugin_name, **arguments)
  File "/home/travis/build/learningequality/kolibri/kolibri/utils/cli.py", line 209, in plugin
    raise PluginDoesNotExist("Plugin does not exist")
kolibri.utils.cli.PluginDoesNotExist: Plugin does not exist
ERROR: InvocationError: '/home/travis/build/learningequality/kolibri/.tox/py2.7/bin/kolibri plugin kolibri.plugins.learn enable'
```

File: kolibri/content/models.py [hunk 3]
```diff
@@ -144,102 +136,51 @@ def get_url(self):
...
-    def clean(self, *args, **kwargs):
-        # self reference exception
-        if self.target_node == self.prerequisite:
-            raise IntegrityError('Cannot self reference as prerequisite.')
-        # immediate cyclic exception
-        elif PrerequisiteContentRelationship.objects.using(self._state.db)\
-                .filter(target_node=self.prerequisite, prerequisite=self.target_node):
-            raise IntegrityError(
-                'Note: Prerequisite relationship is directional! %s and %s cannot be prerequisite of each other!'
-                % (self.target_node, self.prerequisite))
```
Comment (66eli77) [inline kolibri/content/models.py:186]: these rules will not be reinforced at implementation level, but I'm ok with that.

Comment (jamalex) [inline kolibri/content/models.py:186]: Yeah, I was assuming they're already being enforced on the content curation server -- relationships aren't being created locally in the Kolibri app, so no need to enforce here.

File: kolibri/content/urls.py [hunk 4]
```diff
@@ -25,7 +23,7 @@ class ContentNodeFilter(filters.FilterSet):
 
     class Meta:
         model = models.ContentNode
-        fields = ['parent', 'search']
+        fields = ['parent', 'search', 'prerequisite_for', 'has_prerequisite', 'related']
```
Comment (66eli77) [inline kolibri/content/urls.py:26]: 'search' is not needed anymore

Comment (jamalex) [inline kolibri/content/urls.py:26]: How so?

Comment (66eli77) [inline kolibri/content/urls.py:26]: you can use search without specify again in fields

**Comment by 66eli77** [comment]
good to merge, any objections?

File: kolibri/content/serializers.py [hunk 5]
```diff
@@ -96,35 +46,23 @@ def __init__(self, *args, **kwargs):
         super(ContentNodeSerializer, self).__init__(*args, **kwargs)
 
         # enable dynamic fields specification!
-        if not isinstance(self.context['request'], WSGIRequest):
```
Comment (66eli77) [inline kolibri/content/serializers.py:99]: why delete this check? will get this error `'WSGIRequest' object has no attribute 'query_params'` when render the explore page

**Pull request merged**